### DISCUSSION TITLE
`SigmoidBinaryCrossEntropyLoss` implementation details

### DISCUSSION THREAD

**Discussion by sc12752** [deepjavalibrary/djl#2087] [Q&A] [answered]
Hi,

`SigmoidBinaryCrossEntropyLoss` class has two options: 

- [First option](https://github.com/deepjavalibrary/djl/blob/master/api/src/main/java/ai/djl/training/loss/SigmoidBinaryCrossEntropyLoss.java#L72-L75) is when NN output has been already passed through `sigmoid` and it does look like textbook cross entropy loss.
- [Second option](https://github.com/deepjavalibrary/djl/blob/master/api/src/main/java/ai/djl/training/loss/SigmoidBinaryCrossEntropyLoss.java#L67-L70), where NN output is taken as is without `sigmoid` applied, and it's the one I have troubles understanding. 

Can someone please elaborate on why is there no `sigmoid` application in that branch (I would have expected that given a class name and logic behind it) or why do we have a `relu` and later `softPlus` there? This loss function works for me in practice but I'd like to find out some theory as to why.

Thanks!

**Comment by sc12752** [comment]
Another thing to note is `BinaryAccuracy` evaluator along with `Loss.sigmoidBinaryCrossEntropyLoss` always gives `0.0` accuracy while training. I guess I miss something here, too?

**Reply by sc12752** [reply]
This one was actually a rather stupid error of mine: I was always calling `result.getValidateEvaluation("Accuracy")`, even when using `BinaryAccuracy`. The correct call in the latter case is `result.getValidateEvaluation("BinaryAccuracy")`.

**Comment by sc12752** [comment]
Also asked this at [ai.stackexchange.com](https://ai.stackexchange.com/questions/37520/why-is-sigmoidbinarycrossentropyloss-in-djl-implemented-this-way).

**Comment by zachgk** [comment, marked as answer]
This definition was adopted based on [the one in MXNet](https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/gluon/loss.py#L265). I looked back through the history there and it seems to come from [this PR](https://github.com/apache/incubator-mxnet/pull/9404). That PR links to the [definition from TensorFlow](https://github.com/tensorflow/tensorflow/blob/r1.1/tensorflow/python/ops/nn_impl.py#L100-L136) where it explains the derivation
